File: rust/python/kimsfinance/strategies/volatility.py
"""
Volatility-based trading strategies

These strategies use volatility indicators like ATR, Bollinger Bands, and Keltner Channels
to identify breakouts, reversals, and volatility expansion/contraction.
"""


class ATRBreakoutStrategy:
    """
    ATR-based breakout strategy

    Strategy logic:
    - Buy when price breaks above (previous close + ATR * multiplier)
    - Sell when price breaks below (previous close - ATR * multiplier)

    Parameters:
    - period: ATR calculation period (default: 14)
    - multiplier: ATR multiplier for breakout threshold (default: 2.0)
    """

    def __init__(self, period=14, multiplier=2.0):
        self.period = period
        self.multiplier = multiplier
        self.prev_close = None

    def on_data(self, bar, indicators):
        """Trading logic for ATR breakout"""
        atr = indicators.get(f'atr_{self.period}', 0.0)
        current_close = bar['close']

        if self.prev_close is not None and atr > 0:
            upper_breakout = self.prev_close + (atr * self.multiplier)
            lower_breakout = self.prev_close - (atr * self.multiplier)

            # Bullish breakout
            if current_close > upper_breakout:
                self.prev_close = current_close
                return 'buy'
            # Bearish breakout
            elif current_close < lower_breakout:
                self.prev_close = current_close
                return 'sell'

        self.prev_close = current_close
        return 'hold'

    def get_indicators(self):
        return [f'atr_{self.period}']


class BollingerBreakoutStrategy:
    """
    Bollinger Bands breakout strategy

    Strategy logic:
    - Buy when price closes above upper band (breakout)
    - Sell when price closes below lower band (breakdown)
    - Exit when price returns to middle band

    Parameters:
    - period: Bollinger Bands period (default: 20)
    - std_dev: Standard deviation multiplier (default: 2.0)
    """

    def __init__(self, period=20, std_dev=2.0):
        self.period = period
        self.std_dev = std_dev

    def on_data(self, bar, indicators):
        """Trading logic for Bollinger breakout"""
        # Note: Would need multi-output indicator support
        bb_upper = indicators.get('bb_upper', float('inf'))
        bb_lower = indicators.get('bb_lower', float('-inf'))
        bb_middle = indicators.get('bb_middle', bar['close'])
        price = bar['close']

        # Breakout above upper band
        if price > bb_upper:
            return 'buy'
        # Breakdown below lower band
        elif price < bb_lower:
            return 'sell'
        # Mean reversion to middle band (exit signal)
        # In production, would track position state
        return 'hold'

    def get_indicators(self):
        return ['bb_upper', 'bb_lower', 'bb_middle']


class KeltnerBreakoutStrategy:
    """
    Keltner Channels breakout strategy

    Strategy logic:
    - Buy when price breaks above upper Keltner Channel
    - Sell when price breaks below lower Keltner Channel
    - More robust than Bollinger Bands (uses ATR instead of std dev)

    Parameters:
    - ema_period: EMA period for middle line (default: 20)
    - atr_period: ATR period for channel width (default: 10)
    - multiplier: ATR multiplier for channel width (default: 2.0)
    """

    def __init__(self, ema_period=20, atr_period=10, multiplier=2.0):
        self.ema_period = ema_period
        self.atr_period = atr_period
        self.multiplier = multiplier

    def on_data(self, bar, indicators):
        """Trading logic for Keltner breakout"""
        kc_upper = indicators.get('kc_upper', float('inf'))
        kc_lower = indicators.get('kc_lower', float('-inf'))
        price = bar['close']

        # Breakout above upper channel
        if price > kc_upper:
            return 'buy'
        # Breakdown below lower channel
        elif price < kc_lower:
            return 'sell'
        return 'hold'

    def get_indicators(self):
        return ['kc_upper', 'kc_lower', 'kc_middle']


class VolatilityContractionStrategy:
    """
    Volatility contraction/expansion strategy (Bollinger Squeeze)

    Strategy logic:
    - Detect volatility squeeze when BB bands are narrow (< threshold)
    - Enter on first breakout direction after squeeze
    - Exit when volatility expands significantly

    Parameters:
    - period: Bollinger Bands period (default: 20)
    - std_dev: Standard deviation multiplier (default: 2.0)
    - squeeze_threshold: Band width threshold for squeeze (default: 0.02 = 2%)
    """

    def __init__(self, period=20, std_dev=2.0, squeeze_threshold=0.02):
        self.period = period
        self.std_dev = std_dev
        self.squeeze_threshold = squeeze_threshold
        self.in_squeeze = False
        self.prev_close = None

    def on_data(self, bar, indicators):
        """Trading logic for volatility squeeze"""
        bb_upper = indicators.get('bb_upper', 0.0)
        bb_lower = indicators.get('bb_lower', 0.0)
        bb_middle = indicators.get('bb_middle', bar['close'])
        price = bar['close']

        # Calculate band width as percentage of middle band
        if bb_middle > 0:
            band_width = (bb_upper - bb_lower) / bb_middle
        else:
            return 'hold'

        # Detect squeeze (low volatility)
        if band_width < self.squeeze_threshold:
            self.in_squeeze = True
        elif self.in_squeeze and band_width > self.squeeze_threshold * 1.5:
            # Volatility expanding after squeeze
            if self.prev_close is not None:
                # Trade in direction of breakout
                if price > self.prev_close:
                    self.in_squeeze = False
                    self.prev_close = price
                    return 'buy'
                elif price < self.prev_close:
                    self.in_squeeze = False
                    self.prev_close = price
                    return 'sell'

        self.prev_close = price
        return 'hold'

    def get_indicators(self):
        return ['bb_upper', 'bb_lower', 'bb_middle']

File: rust/src/strategies/volatility.rs
//! Volatility-based trading strategies
//!
//! Strategies that capitalize on volatility expansion, contraction, and breakouts.

use crate::backtest::core::{
    IndicatorConfig, IndicatorValues, OHLCVBar, ParameterGrid, ParameterRange, Signal, Strategy,
};

// ============================================================================
// Bollinger Bands Squeeze Strategy
// ============================================================================

/// Bollinger Bands Squeeze Strategy
///
/// Identifies periods of low volatility (squeeze) followed by breakouts.
/// Buys when price breaks above upper band after squeeze.
/// Sells when price breaks below lower band after squeeze.
///
/// # Default Parameters
/// - Period: 20 (standard Bollinger)
/// - Std Dev: 2.0 (standard deviation multiplier)
/// - Squeeze Threshold: 0.05 (band width as % of middle)
///
/// # Optimization Ranges
/// - Period: 15-30
/// - Std Dev: 1.5-2.5
/// - Squeeze Threshold: 0.03-0.08
///
/// # Market Conditions
/// - Best: Markets alternating between consolidation and breakout
/// - Avoid: Continuously trending markets without consolidation
///
/// # Risk Management
/// - Stop Loss: Lower band (dynamic based on volatility)
/// - Take Profit: 2x bandwidth above entry
#[derive(Debug, Clone)]
pub struct BollingerBandsSqueeze {
    pub period: usize,
    pub std_dev: f64,
    pub squeeze_threshold: f64,
    pub take_profit_mult: f64,
    initial_capital: f64,
    in_squeeze: bool,
}

impl BollingerBandsSqueeze {
    pub fn new(period: usize, std_dev: f64, squeeze_threshold: f64) -> Self {
        Self {
            period,
            std_dev,
            squeeze_threshold,
            take_profit_mult: 2.0,
            initial_capital: 10_000.0,
            in_squeeze: false,
        }
    }
}

impl Default for BollingerBandsSqueeze {
    fn default() -> Self {
        Self::new(20, 2.0, 0.05)
    }
}

impl Strategy for BollingerBandsSqueeze {
    fn on_data(&mut self, bar: &OHLCVBar, indicators: &IndicatorValues) -> Signal {
        let key = format!("bb_{}_{}", self.period, self.std_dev);
        let upper_key = format!("{}_upper", key);
        let middle_key = format!("{}_middle", key);
        let lower_key = format!("{}_lower", key);

        let upper = indicators.get(&upper_key).copied().unwrap_or(bar.high);
        let middle = indicators.get(&middle_key).copied().unwrap_or(bar.close);
        let lower = indicators.get(&lower_key).copied().unwrap_or(bar.low);

        let bandwidth = if middle > 0.0 {
            (upper - lower) / middle
        } else {
            0.0
        };

        if bandwidth < self.squeeze_threshold {
            self.in_squeeze = true;
        }

        if self.in_squeeze && bar.close > upper {
            self.in_squeeze = false;
            return Signal::Buy;
        } else if self.in_squeeze && bar.close < lower {
            self.in_squeeze = false;
            return Signal::Sell;
        }

        Signal::Hold
    }

    fn indicators(&self) -> Vec<IndicatorConfig> {
        vec![IndicatorConfig::BollingerBands {
            period: self.period,
            std_dev: self.std_dev,
        }]
    }

    fn parameters(&self) -> ParameterGrid {
        let mut grid = ParameterGrid::new();
        grid.add_range(
            "period",
            ParameterRange::Int { min: 15, max: 30, step: 5 },
        );
        grid.add_range(
            "std_dev",
            ParameterRange::Float { min: 1.5, max: 2.5, step: 0.25 },
        );
        grid.add_range(
            "squeeze_threshold",
            ParameterRange::Float { min: 0.03, max: 0.08, step: 0.01 },
        );
        grid
    }

    fn initial_capital(&self) -> f64 {
        self.initial_capital
    }
}

// ============================================================================
// Bollinger Bands Expansion Strategy
// ============================================================================

/// Bollinger Bands Expansion Strategy (Mean Reversion)
///
/// Fades extreme moves by buying at lower band and selling at upper band.
/// Assumes price will revert to the mean after extreme volatility.
///
/// # Default Parameters
/// - Period: 20 (standard Bollinger)
/// - Std Dev: 2.0 (standard deviation multiplier)
/// - Exit at Middle: true (exit when price returns to middle band)
///
/// # Optimization Ranges
/// - Period: 15-30
/// - Std Dev: 1.5-2.5
///
/// # Market Conditions
/// - Best: Range-bound markets with mean-reverting behavior
/// - Avoid: Strong trending markets (bands will keep expanding)
///
/// # Risk Management
/// - Stop Loss: 3% beyond band (price continues trend)
/// - Take Profit: Middle band (mean reversion)
#[derive(Debug, Clone)]
pub struct BollingerBandsExpansion {
    pub period: usize,
    pub std_dev: f64,
    pub exit_at_middle: bool,
    pub stop_loss_pct: f64,
    initial_capital: f64,
}

impl BollingerBandsExpansion {
    pub fn new(period: usize, std_dev: f64, exit_at_middle: bool) -> Self {
        Self {
            period,
            std_dev,
            exit_at_middle,
            stop_loss_pct: 0.03,
            initial_capital: 10_000.0,
        }
    }
}

impl Default for BollingerBandsExpansion {
    fn default() -> Self {
        Self::new(20, 2.0, true)
    }
}

impl Strategy for BollingerBandsExpansion {
    fn on_data(&mut self, bar: &OHLCVBar, indicators: &IndicatorValues) -> Signal {
        let key = format!("bb_{}_{}", self.period, self.std_dev);
        let upper_key = format!("{}_upper", key);
        let middle_key = format!("{}_middle", key);
        let lower_key = format!("{}_lower", key);

        let upper = indicators.get(&upper_key).copied().unwrap_or(bar.high);
        let middle = indicators.get(&middle_key).copied().unwrap_or(bar.close);
        let lower = indicators.get(&lower_key).copied().unwrap_or(bar.low);

        if bar.close <= lower {
            Signal::Buy
        } else if bar.close >= upper {
            Signal::Sell
        } else if self.exit_at_middle && (bar.close >= middle * 0.995 && bar.close <= middle * 1.005) {
            Signal::Sell
        } else {
            Signal::Hold
        }
    }

    fn indicators(&self) -> Vec<IndicatorConfig> {
        vec![IndicatorConfig::BollingerBands {
            period: self.period,
            std_dev: self.std_dev,
        }]
    }

    fn parameters(&self) -> ParameterGrid {
        let mut grid = ParameterGrid::new();
        grid.add_range(
            "period",
            ParameterRange::Int { min: 15, max: 30, step: 5 },
        );
        grid.add_range(
            "std_dev",
            ParameterRange::Float { min: 1.5, max: 2.5, step: 0.25 },
        );
        grid
    }

    fn initial_capital(&self) -> f64 {
        self.initial_capital
    }
}

// ============================================================================
// ATR Volatility Breakout Strategy
// ============================================================================

/// ATR Volatility Breakout Strategy
///
/// Uses Average True Range (ATR) to identify volatility expansion.
/// Buys when price moves more than N×ATR above previous close.
/// Sells when price moves more than N×ATR below previous close.
///
/// # Default Parameters
/// - ATR Period: 14 (standard ATR)
/// - Breakout Multiplier: 2.0 (2× ATR movement)
/// - Min ATR: 0.5% (avoid low volatility trades)
///
/// # Optimization Ranges
/// - ATR Period: 10-20
/// - Breakout Multiplier: 1.5-3.0
/// - Min ATR: 0.3-1.0%
///
/// # Market Conditions
/// - Best: Volatile markets with clear directional moves
/// - Avoid: Low volatility, choppy markets
///
/// # Risk Management
/// - Stop Loss: 1× ATR below entry
/// - Take Profit: 3× ATR above entry
#[derive(Debug, Clone)]
pub struct ATRVolatilityBreakout {
    pub atr_period: usize,
    pub breakout_multiplier: f64,
    pub min_atr_pct: f64,
    pub stop_loss_atr_mult: f64,
    pub take_profit_atr_mult: f64,
    initial_capital: f64,
    prev_close: f64,
}

impl ATRVolatilityBreakout {
    pub fn new(atr_period: usize, breakout_multiplier: f64, min_atr_pct: f64) -> Self {
        Self {
            atr_period,
            breakout_multiplier,
            min_atr_pct,
            stop_loss_atr_mult: 1.0,
            take_profit_atr_mult: 3.0,
            initial_capital: 10_000.0,
            prev_close: 0.0,
        }
    }
}

impl Default for ATRVolatilityBreakout {
    fn default() -> Self {
        Self::new(14, 2.0, 0.005)
    }
}

impl Strategy for ATRVolatilityBreakout {
    fn on_data(&mut self, bar: &OHLCVBar, indicators: &IndicatorValues) -> Signal {
        let key = format!("atr_{}", self.atr_period);
        let atr = indicators.get(&key).copied().unwrap_or(0.0);

        if self.prev_close == 0.0 {
            self.prev_close = bar.close;
            return Signal::Hold;
        }

        let atr_pct = if self.prev_close > 0.0 {
            atr / self.prev_close
        } else {
            0.0
        };

        if atr_pct < self.min_atr_pct {
            self.prev_close = bar.close;
            return Signal::Hold;
        }

        let breakout_threshold = atr * self.breakout_multiplier;
        let price_move = bar.close - self.prev_close;

        let result = if price_move > breakout_threshold {
            Signal::Buy
        } else if price_move < -breakout_threshold {
            Signal::Sell
        } else {
            Signal::Hold
        };

        self.prev_close = bar.close;

        result
    }

    fn indicators(&self) -> Vec<IndicatorConfig> {
        vec![IndicatorConfig::ATR {
            period: self.atr_period,
        }]
    }

    fn parameters(&self) -> ParameterGrid {
        let mut grid = ParameterGrid::new();
        grid.add_range(
            "atr_period",
            ParameterRange::Int { min: 10, max: 20, step: 2 },
        );
        grid.add_range(
            "breakout_multiplier",
            ParameterRange::Float { min: 1.5, max: 3.0, step: 0.5 },
        );
        grid.add_range(
            "min_atr_pct",
            ParameterRange::Float { min: 0.003, max: 0.010, step: 0.001 },
        );
        grid
    }

    fn initial_capital(&self) -> f64 {
        self.initial_capital
    }
}
